### ISSUE TITLE
L1 partial-range T.copy 在 NZ layout 下 Z-block 交织模式计算错误

### ISSUE THREAD

**Issue by fengz72** [tile-ai/tilelang-ascend#1155]
# Bug: NZ Layout L1 Buffer 部分范围 T.copy 导致 gemm_v0 结果严重错误（地址计算错位）

## 现象

对同一个 NZ layout 的 L1 buffer 进行两次部分范围 `T.copy`（如 `k_l1[0:64, :]` + `k_l1[64:128, :]`），后续 `T.gemm_v0` 输出完全错误（max_diff ≈ 50）。而单次完整范围 `T.copy`（`k_l1[:, :]`）结果正确（max_diff ≈ 0）。

该问题阻碍了 Flash Attention 中 prefix-cache / paged-KV 的 L1 直接拼接优化：将缓存前缀 token 和活跃 token 直接拼入 L1（`k_l1[0:cache_len]` + `k_l1[cache_len:N]`）会破坏 `mma/gemm_v0` 所需的数据布局。

## 根因

昇腾 NPU 中，NZ layout 的 L1 buffer 按 Z-block 交织顺序存储数据，交织范围覆盖 **buffer 的完整 N 维度**（如 128 行）。`copy_gm_to_l1` DMA 引擎根据 **本次拷贝的行数**（runtime `rows` 参数）施加 Z-block 交织，而非 **buffer 的声明 N 维度**。

当 128 行 NZ buffer 通过两次独立的 64 行拷贝填充时：

1. **第一次拷贝**（`rows=64`）：Z-block 仅在行 0–63 之间交织
2. **第二次拷贝**（`rows=64`，偏移 `k_l1[8192]`）：Z-block 仅在行 64–127 之间交织，**如同它们是一个独立 64 行 buffer 的行 0–63**

两次 64 行 NZ-block 传输产生的 **物理布局 ≠ 一次 128 行 NZ-block 传输的物理布局**。`gemm_v0` 按 128 行 NZ 去交织读取 `k_l1[0]` 时，物理 Z-block 位置不匹配，导致数据错乱。

此外，偏移量 `k_l1[8192]`（64×128）按线性布局计算，但 NZ layout buffer 中逻辑行 64 的实际物理偏移可能因 Z-block padding 而不同。

## AscendC 源码对比

**正确（单次完整拷贝）：**
```cpp
tl::ascend::copy_gm_to_l1<bfloat16_t, 128, 128>(
    k_l1[0], K_full[...], 128, 128, 128);
// 单次 DMA：NZ 交织覆盖全部 128 行 → 物理布局正确
```

**错误（两次部分拷贝）：**
```cpp
tl::ascend::copy_gm_to_l1<bfloat16_t, 128, 128>(
    k_l1[0], K_src0[...], 128, 64, 128);      // rows=64
AscendC::PipeBarrier<PIPE_MTE2>();
tl::ascend::copy_gm_to_l1<bfloat16_t, 128, 128>(
    k_l1[8192], K_src1[...], 128, 64, 128);    // rows=64, flat offset 64*128
// 两次 DMA：每次仅对 64 行施加 NZ 交织 → 物理布局错误
```

关键差异：运行参数 `128, 128, 128`（完整）vs `128, 64, 128`（部分）。模板参数 `<128, 128>` 相同，但 `rows=64` 时的 NZ 交织模式与 `rows=128` 时不同。

## 最小复现代码

```python
import tilelang
from tilelang import language as T
from tilelang.intrinsics import make_zn_layout, make_nz_layout
import torch

torch.set_default_device("npu")
torch.manual_seed(0)
tilelang.disable_cache()

BLOCK_M, BLOCK_N, DIM = 64, 128, 128

DEV_CONFIGS = {
    tilelang.PassConfigKey.TL_ASCEND_AUTO_CV_COMBINE: True,
    tilelang.PassConfigKey.TL_ASCEND_AUTO_CV_SYNC: True,
    tilelang.PassConfigKey.TL_ASCEND_AUTO_SYNC: True,
    tilelang.PassConfigKey.TL_ASCEND_MEMORY_PLANNING: True,
}

@tilelang.jit(pass_configs=DEV_CONFIGS)
def full_copy_gemm(dim, block_M=64, block_N=128):
    dtype = "bfloat16"
    accum_dtype = "float32"
    batch = T.symbolic("batch")
    @T.prim_func
    def main(
        Q: T.Tensor([batch, block_M, dim], dtype),
        K_full: T.Tensor([batch, block_N, dim], dtype),
        S_out: T.Tensor([batch, block_M, block_N], accum_dtype),
    ):
        with T.Kernel(1, is_npu=True) as (cid, vid):
            q_l1 = T.alloc_L1([block_M, dim], dtype)
            k_l1 = T.alloc_L1([block_N, dim], dtype)
            l0c = T.alloc_L0C([block_M, block_N], accum_dtype)
            T.annotate_layout({
                q_l1: make_zn_layout(q_l1),
                k_l1: make_nz_layout(k_l1),
            })
            for b_i in T.serial(batch):
                T.copy(Q[b_i, :, :], q_l1[:, :])
                T.copy(K_full[b_i, :, :], k_l1[:, :])          # 完整范围：正确
                T.gemm_v0(q_l1, k_l1, l0c, transpose_B=True, init=True)
                T.copy(l0c, S_out[b_i, :, :])
    return main

@tilelang.jit(pass_configs=DEV_CONFIGS)
def splice_copy_gemm(dim, block_M=64, block_N=128, split=64):
    dtype = "bfloat16"
    accum_dtype = "float32"
    batch = T.symbolic("batch")
    @T.prim_func
    def main(
        Q: T.Tensor([batch, block_M, dim], dtype),
        K_src0: T.Tensor([batch, split, dim], dtype),
        K_src1: T.Tensor([batch, block_N - split, dim], dtype),
        S_out: T.Tensor([batch, block_M, block_N], accum_dtype),
    ):
        with T.Kernel(1, is_npu=True) as (cid, vid):
            q_l1 = T.alloc_L1([block_M, dim], dtype)
            k_l1 = T.alloc_L1([block_N, dim], dtype)
            l0c = T.alloc_L0C([block_M, block_N], accum_dtype)
            T.annotate_layout({
                q_l1: make_zn_layout(q_l1),
                k_l1: make_nz_layout(k_l1),
            })
            for b_i in T.serial(batch):
                T.copy(Q[b_i, :, :], q_l1[:, :])
                T.copy(K_src0[b_i, :, :], k_l1[0:split, :])    # 部分范围：错误
                T.copy(K_src1[b_i, :, :], k_l1[split:block_N, :]) # 部分范围：错误
                T.gemm_v0(q_l1, k_l1, l0c, transpose_B=True, init=True)
                T.copy(l0c, S_out[b_i, :, :])
    return main

# 验证
q = torch.randn(1, BLOCK_M, DIM).bfloat16()
k_full = torch.randn(1, BLOCK_N, DIM).bfloat16()
k_src0 = k_full[:, 0:64, :]
k_src1 = k_full[:, 64:, :]
ref = torch.einsum("qd,kd->qk", q[0].float(), k_full[0].float())

func_full = full_copy_gemm(dim=DIM)
s_full = torch.zeros(1, BLOCK_M, BLOCK_N, dtype=torch.float32)
func_full(q.npu(), k_full.npu(), s_full.npu())

func_splice = splice_copy_gemm(dim=DIM, split=64)
s_splice = torch.zeros(1, BLOCK_M, BLOCK_N, dtype=torch.float32)
func_splice(q.npu(), k_src0.npu(), k_src1.npu(), s_splice.npu())

torch.npu.synchronize()
print(f"Full copy   max_diff: {(s_full[0] - ref).abs().max():.6f}")   # ≈ 0
print(f"Splice copy max_diff: {(s_splice[0] - ref).abs().max():.2f}") # ≈ 50
```

## 期望输出

```
Full copy   max_diff: 0.000008
Splice copy max_diff: 49.62
```

**Comment by ArmandAlbert** [comment]
PR #1243  has fixed. Please check.
Testcases:
<html>
<body>
<html><head>

Group | Scenario | Operator | Layout / Transpose | Variable Type | Number of Cases
-- | -- | -- | -- | -- | --
G1 | Full Block Copy Baseline | full_copy_annotated / full_copy_plain | zN+nZ / default zN; transB=T/F | Constant | 4
G2 | K-tail Zeroing Regression | full_copy_plain | Default zN; transB=F | Constant | 3
G3 | 2-way nZ splice | splice_2way_nz | nZ; transB=T | Constant | 12
G4 | 3-way splice + Unfilled Tail | splice_3way_tail | nZ; transB=T | Constant | 4
G5 | 2-way zN splice | splice_2way_zn | zN; transB=F | Constant | 3
G6 | Combined splice + K-tail | splice_with_k_tail | nZ; transB=T | Constant | 2
G7 | Dynamic Batch + Dynamic Split | splice_2way_nz_dynamic_batch | nZ; transB=T | Symbolic batch + split | 12
G8 | Dynamic K K-tail | full_copy_plain_dynamic_k | Default zN; transB=F | Symbolic K | 3
G9 | Dynamic split splice | splice_2way_nz_dynamic_split | nZ; transB=T | Symbolic split | 6
